rust/src/io/commit.rs

```diff
@@ -0,0 +1,406 @@
+        manifest_writer: ManifestWriter,
+    ) -> Result<(), CommitError>;
+}
+
+/// Errors that can occur when committing a manifest.
+pub enum CommitError {
+    /// Another transaction has already been written to the path
+    CommitConflict,
+    /// Something else went wrong
+    OtherError(crate::Error),
+}
+
+impl From<crate::Error> for CommitError {
+    fn from(e: crate::Error) -> Self {
+        Self::OtherError(e)
+    }
+}
+
+impl From<CommitError> for crate::Error {
+    fn from(e: CommitError) -> Self {
+        match e {
+            CommitError::CommitConflict => Self::Internal {
+                message: "Commit conflict".to_string(),
+            },
+            CommitError::OtherError(e) => e,
+        }
+    }
+}
+
+/// Whether we have issued a warning about using the unsafe commit handler.
+static WARNED_ON_UNSAFE_COMMIT: AtomicBool = AtomicBool::new(false);
+
+/// A naive commit implementation that does not prevent conflicting writes.
+///
+/// This will log a warning the first time it is used.
+pub struct UnsafeCommitHandler;
+
+#[async_trait::async_trait]
+impl CommitHandler for UnsafeCommitHandler {
+    async fn commit(
+        &self,
+        manifest: &mut Manifest,
+        indices: Option<Vec<Index>>,
+        path: &Path,
+        object_store: &ObjectStore,
+        manifest_writer: ManifestWriter,
+    ) -> Result<(), CommitError> {
+        // Log a one-time warning
+        if !WARNED_ON_UNSAFE_COMMIT.load(std::sync::atomic::Ordering::Relaxed) {
+            WARNED_ON_UNSAFE_COMMIT.store(true, std::sync::atomic::Ordering::Relaxed);
+            log::warn!(
+                "Using unsafe commit handler. Concurrent writes may result in data loss. \
+                 Consider providing a commit handler that prevents conflicting writes."
+            );
+        }
+
+        // Write the manifest naively
+        manifest_writer(object_store, manifest, indices, path).await?;
+
+        Ok(())
+    }
+}
+
+impl Debug for UnsafeCommitHandler {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("UnsafeCommitHandler").finish()
+    }
+}
+
+/// A commit implementation that uses a temporary path and renames the object.
+///
+/// This only works for object stores that support atomic rename if not exist.
+pub struct RenameCommitHandler;
+
+#[async_trait::async_trait]
+impl CommitHandler for RenameCommitHandler {
+    async fn commit(
+        &self,
+        manifest: &mut Manifest,
+        indices: Option<Vec<Index>>,
+        path: &Path,
+        object_store: &ObjectStore,
+        manifest_writer: ManifestWriter,
+    ) -> Result<(), CommitError> {
+        // Create a temporary object, then use `rename_if_not_exists` to commit.
+        // If failed, clean up the temporary object.
+
+        // Add .tmp_ prefix to the path
+        let mut parts: Vec<_> = path.parts().collect();
+        let new_name = format!(".tmp_{}", parts.last().unwrap().as_ref());
```

Comment:
What's to prevent a concurrent writer from writing the same temporary object? We should inject a UUID into the path to avoid conflicts.